tests/test_page.py

```diff
@@ -10,23 +10,48 @@
 
 
 def test_base_url(base_url, page):
-    assert base_url == page.url
+    assert base_url == page.canonical_url
 
 
-def test_url_keywords(base_url, selenium):
+def test_canonical_url_absolute(base_url, selenium):
+    url_template = 'https://www.test.com/'
+
+    class MyPage(Page):
+        URL_TEMPLATE = url_template
+    page = MyPage(selenium, base_url)
+    assert url_template == page.canonical_url
+
+
+def test_canonical_url_empty(selenium):
+    page = Page(selenium)
+    assert page.canonical_url == ''
+
+
+def test_canonical_url_keywords(base_url, selenium):
     value = str(random.random())
 
     class MyPage(Page):
-        _url = '{key}'
-    page = MyPage('foo', selenium, key=value)
-    assert value == page.url
+        URL_TEMPLATE = '{key}'
+    page = MyPage(selenium, base_url, key=value)
+    assert base_url + value == page.canonical_url
+
+
+def test_canonical_url_prepend(base_url, page, selenium):
+    assert base_url == page.canonical_url
```

Comment:
I don't think this is testing that the canonical URL is being appended to the base URL. It looks like the URL_TEMPLATE is empty and so we're just returning the base URL. To test concatenation we'd need a path for URL_TEMPLATE.